x-pack/platform/plugins/shared/onechat/public/application/context/send_message/use_subscribe_to_chat_events.ts

```diff
@@ -35,97 +36,119 @@ export const useSubscribeToChatEvents = ({
+            }
+            // full message received, override chunk buffer
+            else if (isMessageCompleteEvent(event)) {
+              conversationActions.setAssistantMessage({
+                assistantMessage: event.data.message_content,
+              });
+            } else if (isToolProgressEvent(event)) {
+              conversationActions.setToolCallProgress({
+                progress: { message: event.data.message },
+                toolCallId: event.data.tool_call_id,
+              });
+              // Individual tool progression message should also be displayed as reasoning
+              setAgentReasoning(event.data.message);
+            } else if (isReasoningEvent(event)) {
+              conversationActions.addReasoningStep({
+                step: createReasoningStep({
+                  reasoning: event.data.reasoning,
+                  transient: event.data.transient,
+                }),
+              });
+              setAgentReasoning(event.data.reasoning);
+            } else if (isToolCallEvent(event)) {
+              conversationActions.addToolCall({
+                step: createToolCallStep({
+                  params: event.data.params,
+                  results: [],
+                  tool_call_id: event.data.tool_call_id,
+                  tool_id: event.data.tool_id,
+                }),
+              });
+            } else if (isBrowserToolCallEvent(event)) {
+              // Check if this is a browser tool call and execute it immediately
+              const toolId = event.data.tool_id;
+              if (toolId && browserToolExecutor && browserApiTools) {
+                const toolDef = browserApiTools.find((tool) => tool.id === toolId);
+                if (toolDef) {
+                  const toolsMap = new Map([[toolId, toolDef]]);
+                  browserToolExecutor
+                    .executeToolCalls(
+                      [
+                        {
+                          tool_id: toolId,
+                          call_id: event.data.tool_call_id,
+                          params: event.data.params,
+                          timestamp: Date.now(),
+                        },
+                      ],
+                      toolsMap
+                    )
+                    .catch((error) => {
+                      // eslint-disable-next-line no-console
+                      console.error('Failed to execute browser tool:', error);
+                    });
+                }
               }
+            } else if (isToolResultEvent(event)) {
+              const { tool_call_id: toolCallId, results } = event.data;
+              conversationActions.setToolCallResult({ results, toolCallId });
+            } else if (isRoundCompleteEvent(event)) {
+              // Now we have the full response and can stop the loading indicators
+              setIsResponseLoading(false);
+            } else if (isConversationCreatedEvent(event)) {
```

Comment:
You should be able to remove this now, I'm fine with keeping it if you'd like to keep it for clarity that this event exists. I'm also not sure of the timing between `isRoundCompleteEvent` and `onSettled` running, so it might be better to `setIsResponseLoading(false)` here so that the user gets that feedback sooner.